src/metabase/lib/join.cljc

```diff
@@ -751,67 +752,75 @@
-    _rhs-column-or-nil :- [:maybe lib.metadata/ColumnMetadata]]
+    _lhs-column-or-nil :- [:maybe ::lib.schema.metadata/column]
+    _rhs-column-or-nil :- [:maybe ::lib.schema.metadata/column]]
    ;; currently hardcoded to these six operators regardless of LHS and RHS.
    lib.filter.operator/join-operators))
 
-(mu/defn ^:private pk-column :- [:maybe lib.metadata/ColumnMetadata]
-  "Given something `x` (e.g. a Table metadata) find the PK column."
+(mu/defn ^:private fk-columns-to :- [:maybe [:sequential
+                                             {:min 1}
+                                             [:and
+                                              ::lib.schema.metadata/column
+                                              [:map
+                                               [::target ::lib.schema.metadata/column]]]]]
+  "Find FK columns in `source` pointing at a column in `target`. Includes the target column under the `::target` key."
   [query        :- ::lib.schema/query
    stage-number :- :int
-   x
-   options]
-  (m/find-first lib.types.isa/primary-key?
-                (lib.metadata.calculation/visible-columns query stage-number x options)))
-
-(mu/defn ^:private fk-column-for-pk-in :- [:maybe lib.metadata/ColumnMetadata]
-  [pk-col          :- [:maybe lib.metadata/ColumnMetadata]
-   visible-columns :- [:maybe [:sequential lib.metadata/ColumnMetadata]]]
-  (when-let [pk-id (:id pk-col)]
-    (let [candidates (filter (fn [{:keys [fk-target-field-id], :as col}]
-                               (and (lib.types.isa/foreign-key? col)
-                                    (= fk-target-field-id pk-id)))
-                             visible-columns)
-          primary (first candidates)]
-      (if (some #(= (:id %) (:id primary)) (rest candidates))
-        (when (not-any? #(= (:lib/desired-column-alias %) (:lib/desired-column-alias primary)) (rest candidates))
-          ;; there are multiple candidates but :lib/desired-column-alias disambiguates them,
-          ;; so reference by name
-          (dissoc primary :id))
-        primary))))
-
-(mu/defn ^:private fk-column-for :- [:maybe lib.metadata/ColumnMetadata]
-  "Given a query stage find an FK column that points to the PK `pk-col`."
-  [query        :- ::lib.schema/query
-   stage-number :- :int
-   pk-col       :- [:maybe lib.metadata/ColumnMetadata]]
-  (when pk-col
-    (let [visible-columns (lib.metadata.calculation/visible-columns query stage-number (lib.util/query-stage query stage-number))]
-      (fk-column-for-pk-in pk-col visible-columns))))
-
-(mu/defn suggested-join-condition :- [:maybe ::lib.schema.expression/boolean] ; i.e., a filter clause
-  "Return a suggested default join condition when constructing a join against `joinable`, e.g. a Table, Saved
-  Question, or another query. A suggested condition will be returned if the query stage has a foreign key to the
+   source
+   target]
+  (let [target-columns (delay
```

Comment:
Why delay here? Should we be wary of `visible-columns` calls elsewhere?

Reply:
mostly just to avoid calculating it if unneeded. Calculating it isn't free, so it's just a performance optimization